ffi/util.lua

```diff
@@ -272,9 +285,18 @@ function util.runInSubProcess(func, with_pipe)
         os.exit(0)
     end
     -- parent/main process
-    if pid == -1 then -- on failure, fork() returns -1
+    if opid < 0 then -- on failure, fork() returns -1
```

Comment:
We never get the inner pid in the main process right?
So, why `opid` here? Except for clarity. But it's a bit confusing (reading it for the first time) that you use `opid` in a non double_fork `if` - and that at the end, we return `pid`. May be use just one of `opid` or `pid` everywhere after that, and mention it has no meaning when double_fork ?

Otherwise, looks fine and _not too convoluted_ enough :)

Reply:
Unless I'm having C hangups, I can't declare a local inside a branch, that'd limit it to that branch's scope. So I can't call it pid if !double_fork and opid if double_fork ;).

So I *have* to set opid in the function scope, which means it gets to be used even when !double_fork, in which case opid == pid ;).

Reply:
And I do have to make the distinction between inner and outer fork in that final check.

Reply:
Oh, wait, I get what you mean. -_-".

Reply:
Not really following each case you mention, because it feels I don't have to, if the following is right:

you could just use the single local variable pid defined outside the branch and forks.
Then in each branch, you're a different forked process: each forked process can use the single local pid the way it wants too: it won't be seen by the other forked processes as each one gets its own copy of the memory once it writes to: there shouldn't be any automatically shared memory between all these forked processes and the main one. So, the parent will just see the one from the first C.fork().

Reply:
Yeah, making it exceedingly clear that pid != pid depending on whether you're in the child process or the parent process would make things more convoluted, and what I went with was indeed not very clear, thanks ;).

Reply:
May be mention that in the double_fork case, the pid we return is of no use and not valid (as it has already been wait()ed).
Dunno if we should return pid=nil, or that first fork pid anyway (so the caller can have something to wait() on, that will return the expected value of _no more there_.

Reply:
Yeah, good point.

I guess nil sounds good, because I'm not quite sure what waitpid returns for an already-reaped PID (ECHILD?).

Reply:
Yep, ECHILD it is.

Reply:
Okay, just documented it, since isSUbprocessDone can handle it sanely. Returning nil would be potentially less friendly ;).